### PR TITLE
Add iot-simplified-installer and iot-qcow2 image types to Fedora

### PR DESCRIPTION
**Pull request by achilleas-k**
_Migration of https://github.com/osbuild/osbuild-composer/pull/3282._

Adds new image types to Fedora:
- iot-simplified-installer (F38+)
- iot-qcow2-image

This PR also introduces a new interface, `FilePipeline`, which is implemented by pipelines that produce a single file (typically an image file, but it could be any single file).  Pipelines that work on single files (like `XZ`, `QCOW2`, `VMDK`) can now take any `FilePipeline` as input.

### CODE REVIEW

File: pkg/distro/fedora/package_sets.go
Comment (teg) [inline pkg/distro/fedora/package_sets.go:180]: For a follow-up: this kind of logic would be perfect to have in the Platform itself.

File: pkg/distro/fedora/distro.go
Comment (supakeen) [inline pkg/distro/fedora/distro.go:822]: Do we want to enable the firmware overlays and /boot setup for the Pi here?

File: pkg/distro/fedora/images.go
Comment (supakeen) [inline pkg/distro/fedora/images.go:528]: Why doesn't this platform also get the `bpIgnition` that raw gets?

**Comment by supakeen** [comment]
Just some questions happy to approve after :)

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:822]: I based these decisions entirely on the PR in composer that I moved over.  I'd ask @runcom.

File: pkg/distro/fedora/images.go
Comment (achilleas-k) [reply pkg/distro/fedora/images.go:528]: Similar answer: the iot-qcow2 in the original PR didn't get the `ignition.config.url` option: https://github.com/paulwhalen/osbuild-composer/blob/e1d879691eb74bde29746e3dfa20447efad70b35/internal/distro/fedora/images.go#L501
(note that in that PR the image function was completely separate and here it's shared between the two image types).

Perhaps it was an oversight.  I'll defer to @runcom again.

File: pkg/distro/fedora/package_sets.go
Comment (achilleas-k) [reply pkg/distro/fedora/package_sets.go:221]: Good catch.  Yes, let's change them.  It doesn't change the manifests.

**Comment by achilleas-k** [comment]
> Just some questions happy to approve after :)

Confirmed with @nullr0ute that both decisions you pointed out were intentional.

**Review by thozza — DISMISSED** [review]
Very nice work! I like all the refactoring and simplification in this PR.

One thing that we discussed off-channel is the difference between `Pipeline` and `TreePipeline` and that it actually don't contain all methods of the `Pipeline` interface, unlike the `FilePipeline` does...

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:684]: Very good point, yes.  Should be defined on the distro object probably like we do for the other const strings.

File: pkg/distro/fedora/partition_tables.go
Comment (achilleas-k) [reply pkg/distro/fedora/partition_tables.go:281]: True, it's used everywhere.

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:162]: Yeah I was wondering about that.  Whether there's some tooling out there that is using the old aliases and somehow wants to maintain convention.  But that's not the case.

Comment (achilleas-k) [reply pkg/distro/fedora/distro.go]: Removed.

File: pkg/distro/fedora/package_sets.go
Comment (achilleas-k) [reply pkg/distro/fedora/package_sets.go]: Moved to platforms.

File: pkg/distro/fedora/images.go
Comment (achilleas-k) [reply pkg/distro/fedora/images.go:384]: Opened issue to add this in a follow-up.

File: pkg/distro/fedora/partition_tables.go
Comment (achilleas-k) [reply pkg/distro/fedora/partition_tables.go:281]: Opened issue to add this in a follow-up.

**Review by thozza — APPROVED** [review]
Thanks for the changes!

I'm sorry, but I found some additional small things :innocent: . Nevertheless, all of them can be changed in a follow-up :wink:

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:208]: I thought I looked for all the newly added ones but I guess I was sloppy.  Thanks :)

**Comment by achilleas-k** [comment]
Thanks!  I think I caught them all now :)

**Comment by achilleas-k** [comment]
IoT installer failed to build yesterday because of missing `biosdevname`.  Looking into that.

File: pkg/distro/fedora/distro.go
Comment (thozza) [inline pkg/distro/fedora/distro.go:799]: ditto

File: pkg/distro/fedora/distro.go
Comment (thozza) [inline pkg/distro/fedora/distro.go:644]: I got tripped by `grub2` being still here. After looking around, it turns out that there is no `grub2` binary RPM. It is provided by the `grub2-pc` package.

```
[root@15aa9b685f65 /]# dnf provides grub2
Last metadata expiration check: 0:00:15 ago on Wed Aug 16 09:06:47 2023.
grub2-pc-1:2.06-89.fc38.x86_64 : Bootloader with support for Linux, Multiboot, and more
Repo        : fedora
Matched from:
Provide    : grub2 = 1:2.06-89.fc38

grub2-pc-1:2.06-95.fc38.x86_64 : Bootloader with support for Linux, Multiboot, and more
Repo        : updates
Matched from:
Provide    : grub2 = 1:2.06-95.fc38
```

So, it should not need to be listed here. I also checked other distros and we don't list `grub2` in any of them (not in platforms and also not in package sets).

`biosdevname` is also an interesting story... In other distros, we list it in the image package set. I would argue that it is not a firmware, so it probably should not be here. But I'm OK with keeping it as is for now...

**Review by thozza — DISMISSED** [review]

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:644]: `biosdevname` should be added conditionally based on the option in the COI pipeline.

Comment (achilleas-k) [reply pkg/distro/fedora/distro.go]: I did the `biosdevname` conditional and dropped `grub2`.

**Comment by achilleas-k** [comment]
I think we can merge this now and follow-up with the things I put in the issue another time.

**Review by thozza — DISMISSED** [review]
LGTM. There are still some `grub2` packages in a pipeline `FirmwarePackages` definition, but let's do that in a follow-up as part of the issue to sort out these things.

**Review by thozza — APPROVED** [review]

**Comment by achilleas-k** [comment]
The rebase on `main` in the merge-queue was failing because, in this PR, I added image types to the unit tests only for F38 and F39 and then the rebase pulled in F40 and the list was wrong: https://github.com/achilleas-k/images/blob/bd7df48a51fe0ced433e24484a5689456d81bbd1/pkg/distro/fedora/distro_test.go#L559

In hindsight, the way I added those to the tests was a bit weird; I should have used the version range, but it seemed simpler at the time.

Those tests are a bit weird in general though.  For me at least, they're always busy-work, going around all the unit tests and updating static lists, that I need to do after adding an image type.  I suppose they can catch issues if you're refactoring something in a distro and accidentally drop an image type from the registry, but when adding a new one, they sort of do the opposite: if I forget to add a new image type to a distro, I'll probably forget about the unit test lists as well and the tests will pass.  The most common scenario is that I add an image type to the distro, then the unit test fails prompting me to hunt down all the lists.

**Comment by achilleas-k** [comment]
> The rebase on `main` in the merge-queue was failing because, in this PR, I added image types to the unit tests only for F38 and F39 and then the rebase pulled in F40 and the list was wrong: https://github.com/achilleas-k/images/blob/bd7df48a51fe0ced433e24484a5689456d81bbd1/pkg/distro/fedora/distro_test.go#L559
> 
> In hindsight, the way I added those to the tests was a bit weird; I should have used the version range, but it seemed simpler at the time.
> 
> Those tests are a bit weird in general though. For me at least, they're always busy-work, going around all the unit tests and updating static lists, that I need to do after adding an image type. I suppose they can catch issues if you're refactoring something in a distro and accidentally drop an image type from the registry, but when adding a new one, they sort of do the opposite: if I forget to add a new image type to a distro, I'll probably forget about the unit test lists as well and the tests will pass. The most common scenario is that I add an image type to the distro, then the unit test fails prompting me to hunt down all the lists.

I think we can replace this with a static list outside the code and a run of `./cmd/list-images` to compare against.  It would be like the manifest checks in composer, only it's just a list of repos x arches x image types and it would take a second to run in the unit tests.

**Pull request merged**
